File: frontend/src/app/types/occupation.ts
import { DoseBand } from './dose';

export type OccupationStatus = 'occupied' | 'retained' | 'released';
export type OccupationReleaseReason = 'review_rejected' | 'plan_archived' | '';

export interface BudgetOccupation {
  id: number;
  worker_id: number;
  worker_code: string;
  worker_name: string;
  plan_id: number;
  plan_code: string;
  assessment_id: number;
  occupation_status: OccupationStatus;
  dose_msv: number;
  period_dose_msv: number;
  occupied_by: number;
  occupied_at: string;
  retained_by?: number;
  retained_at?: string;
  released_by?: number;
  released_at?: string;
  release_reason: OccupationReleaseReason;
  version: number;
  planning_only: boolean;
  automatic_work_permit: boolean;
}

export interface WorkerBudgetBalance {
  worker_id: number;
  worker_code: string;
  worker_name: string;
  authorization_level: string;
  profile_status: string;
  period_start: string;
  administrative_limit_msv: number;
  annual_legal_limit_msv: number;
  verified_dose_msv: number;
  active_occupation_msv: number;
  committed_dose_msv: number;
  remaining_admin_msv: number;
  remaining_legal_msv: number;
  active_occupation_count: number;
  active_occupation_ids: number[];
  risk_band: DoseBand;
  requires_manual_review: boolean;
  escalation_explanation: string;
  threshold_version: string;
  planning_only: boolean;
  automatic_work_permit: boolean;
}

File: backend/internal/dosebudget/occupation.go
package dosebudget

import "fmt"

// OccupationInput carries the immutable facts needed to turn one submitted
// assessment into one budget occupation.
type OccupationInput struct {
	WorkerID      uint
	PlanID        uint
	AssessmentID  uint
	DoseMSV       float64
	PeriodDoseMSV float64
}

// Validate checks that an occupation never carries non-finite or negative
// dose. A zero increment is allowed because genuine zero-dose planning
// scenarios still need their one-time occupation evidence row.
func (input OccupationInput) Validate() error {
	if input.WorkerID == 0 || input.PlanID == 0 || input.AssessmentID == 0 {
		return fmt.Errorf("%w: occupation requires worker, plan and assessment references", ErrInvalidDoseInput)
	}
	if !finite(input.DoseMSV) || input.DoseMSV < 0 {
		return fmt.Errorf("%w: occupied dose must be finite and non-negative", ErrInvalidDoseInput)
	}
	if !finite(input.PeriodDoseMSV) || input.PeriodDoseMSV < 0 {
		return fmt.Errorf("%w: period dose must be finite and non-negative", ErrInvalidDoseInput)
	}
	return nil
}

// PlannedDose is the single formula used when a submission occupies budget:
// estimated_rate_msvh × planned_minutes ÷ 60. It mirrors CalculateProjection
// but does not require a current period dose.
func PlannedDose(estimatedRateMSVH float64, plannedMinutes int) (float64, error) {
	projection, err := CalculateProjection(0, estimatedRateMSVH, plannedMinutes)
	if err != nil {
		return 0, err
	}
	return projection.PlannedDoseMSV, nil
}

// WorkerBudgetBalance is one worker's budget view: verified period dose,
// active occupations and remaining quota in a single structure.
type WorkerBudgetBalance struct {
	WorkerID               uint
	AdministrativeLimitMSV float64
	LegalLimitMSV          float64
	VerifiedDoseMSV        float64
	ActiveOccupationMSV    float64
	CommittedDoseMSV       float64
	RemainingAdminMSV      float64
	RemainingLegalMSV      float64
	ActiveOccupationCount  int
	RiskBand               string
	RequiresManualReview   bool
	EscalationExplanation  string
	PlanningOnly           bool
}

// BuildWorkerBalance assembles the single budget view for one worker.
// Over-limit values never block: the result only flags manual review and
// never represents an authorization to work.
func BuildWorkerBalance(
	workerID uint,
	administrativeLimitMSV, legalLimitMSV, verifiedDoseMSV, activeOccupationMSV float64,
	activeOccupationCount int,
	nearLegalRatio float64,
	thresholdVersion string,
) (WorkerBudgetBalance, error) {
	decision, err := Evaluate(verifiedDoseMSV+activeOccupationMSV, Thresholds{
		AdministrativeLimitMSV: administrativeLimitMSV,
		LegalLimitMSV:          legalLimitMSV,
		NearLegalRatio:         nearLegalRatio,
		Version:                thresholdVersion,
	})
	if err != nil {
		return WorkerBudgetBalance{}, err
	}
	committed := roundDose(verifiedDoseMSV + activeOccupationMSV)
	return WorkerBudgetBalance{
		WorkerID:               workerID,
		AdministrativeLimitMSV: administrativeLimitMSV,
		LegalLimitMSV:          legalLimitMSV,
		VerifiedDoseMSV:        roundDose(verifiedDoseMSV),
		ActiveOccupationMSV:    roundDose(activeOccupationMSV),
		CommittedDoseMSV:       committed,
		RemainingAdminMSV:      decision.RemainingAdminMSV,
		RemainingLegalMSV:      decision.RemainingLegalMSV,
		ActiveOccupationCount:  activeOccupationCount,
		RiskBand:               decision.RiskBand,
		RequiresManualReview:   decision.RequiresManualReview,
		EscalationExplanation:  decision.EscalationExplanation,
		PlanningOnly:           true,
	}, nil
}
